### PR TITLE
[SIP] sqlite host component

### PR DESCRIPTION
**Pull request by rylev**
This introduces [a SIP](https://github.com/fermyon/spin/blob/sqlite-host-component/docs/content/sips/000-sqlite.md) for adding a sqlite interface to Spin as well as an initial MVP implementation.

Much like the key-value interface, this would be a "batteries included" implementation where users gain access to a SQL database with no configuration required. 

I have also created [an example TODO application](https://github.com/rylev/spin-todo) that uses this MVP implementation.

### CODE REVIEW

File: docs/content/sips/000-sqlite.md
Comment (itowlson) [inline docs/content/sips/000-sqlite.md:66]: My impression of a prepared statement was that it generally didn't include parameter values (or, at least, that you could reuse it with different parameter values).  This interface seems to bake the parameter values into the prepared statement.  Is that intentional?

File: docs/content/sips/000-sqlite.md
Comment (itowlson) [inline docs/content/sips/000-sqlite.md:83]: nit: the "type" would be e.g. integer or string; these are values

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md:1]: I definitely think we want to be very clear that the semantics of this interface are sqlite. I think we're asking for trouble if we brand this as generic SQL. This is definitely a point that is causing a lot of discussion about the feasibility of `wasi-sql` (as opposed to `wasi-mysql`, `wasi-postgresql`, etc.)

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md:64]: I agree. I've gone ahead and removed prepared statements because from the host side there is no difference so the added complexity for the guest gets us nothing.

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md:83]: Changed this to just `Value` but happy to keep bikeshedding on it.

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md:72]: Done!

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md:103]: Add this and made a note that we may want to select multiple of these options.

**Comment by rylev** [comment]
@itowlson I think this is ready for another look. The implementation pretty much matches the SIP now. That means the open questions still remain, but we have a working prototype. I guess the question now is: should we answer the open questions or merge and *then* figure out the open questions?

(FYI: the CI failure is because `spin-componentize` needs to be updated)

**Review by itowlson — COMMENTED** [review]
Commented only on the SIP (haven't read through the implementation but can if you want)

File: crates/sqlite/src/lib.rs
Comment (dicej) [inline crates/sqlite/src/lib.rs:140]: Can replace the `async { ... }.await` with a normal block here.

Comment (dicej) [reply crates/sqlite/src/lib.rs:140]: And likewise for the other methods in this `impl` block.

Comment (dicej) [reply crates/sqlite/src/lib.rs:140]: Also, can we wrap these in `tokio::task::block_in_place` like we do in https://github.com/fermyon/spin/blob/b1ac4d8307948bbcd833fc689be3471f8fa174a0/crates/key-value-sqlite/src/lib.rs#L71-L138?

File: docs/content/sips/000-sqlite.md
Comment (dicej) [reply docs/content/sips/000-sqlite.md:23]: For me the question is: do we want to use the opportunity to help _make_ `wasi-sql` ready?

Specifically, I'm thinking it would be nice to follow the precedent we set with key-value such that you can run the same code locally and in Cloud with only a runtime config change.  And since that's kind of the point of `wasi-sql`, we could implement it and making any changes required as necessary.

I'm wary about adding interfaces which can only practically be implemented locally, so I'd either like to see something more like `wasi-sql` or a plan for providing a "cloud-scale" SQLite service.

Comment (rylev) [reply docs/content/sips/000-sqlite.md:23]: To add on top of what Radu says: `wasi-sql` is a long way from stabilization. We can use this work to help inform the evolution of `wasi-sql`, but there are a lot of open questions about the viability of `wasi-sql` in general since sql flavors differ very substantially. Focusing on one sql flavor, sqlite, will make our lives a lot easier and unblock things much more quickly than trying to conform to a future `wasi-sql` standard.

File: Cargo.toml
Comment (rylev) [reply Cargo.toml:132]: Yes, though now that [I have this up on GitHub](https://github.com/rylev/preview1.wasm/tree/sqlite) I can at least refer to the a git dependency and not on a local dependency. But before we merge this we'll want to merge the support into spin-componentize.

File: crates/sqlite/src/lib.rs
Comment (rylev) [reply crates/sqlite/src/lib.rs:140]: Yes though we might move to a different sqlite library which has an async interface.

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md:49]: Well it's not that the name is defined, it's that there is no database by that name. This differs from `access-denied` in that in the `access-denied` case there is explicitly no access granted for a database with that name while in the `no-such-database` case it's not a question of permission, there simply is no database that exists for that identifier. Any suggestions on how to make that clearer?

File: docs/content/sips/000-sqlite.md
Comment (rylev) [reply docs/content/sips/000-sqlite.md]: Changed the wit definitions to not encode the column name for every row. This would be needlessly expensive for large result sets.

File: docs/content/sips/000-sqlite.md
Comment (itowlson) [reply docs/content/sips/000-sqlite.md:49]: "No database named {...} is defined." or "There is no database named {...}." maybe?  (Of these, I somewhat prefer the first, but I am sure there are better options.)

Your comment also made me wonder about which of "not defined" and "not granted access" takes precedence.  If I write a component that tries to use a database called `no-ivans`, but the manifest does not grant my component access to `no-ivans`, will my component get `access-denied` rather than "leaking" information about whether the database exists or doesn't exist?

(Longer term this should be a non-issue because `spin up` should fail manifest validation if it refers to a name that isn't defined, but I don't think we can do that today.)

File: docs/content/sips/000-sqlite.md
Comment (vdice) [inline docs/content/sips/000-sqlite.md:55]: Just checking on the wording here -- Should be 'The database has reached its capacity'?

**Review by vdice — COMMENTED** [review]
Quick proofreading review for the SIP

File: crates/sqlite/src/lib.rs
Comment (rylev) [reply crates/sqlite/src/lib.rs]: So the reason I wrapped these in `async` blocks is the same reason it's done in `key-value` crate. We want the `?` to propagate the inner error of the method's return type. `async` blocks have different scoping rules than regular blocks and allow for a "try block" like control flow. I agree it's a bit ugly, but the alternatives are also not that great...

Comment (rylev) [reply crates/sqlite/src/lib.rs]: We could do something like this: https://github.com/fermyon/spin/compare/sqlite-host-component...block-in-place

**Comment by rylev** [comment]
I am marking this PR as ready to review. While this feature is not complete, I do believe it is useful enough for folks to start experimenting with. Below I've tried to capture the current state and the open questions we need to answer before we can officially make this a stable feature:

This PR provides the following features:
* Support in spin for a sqlite host implementation that provides a default sqlite database when running spin locally.
* Support for runtime configuration for adding additional local sqlite databases.
* Rust SDK support for targeting these databases from a Rust based spin component
* The ability to run arbitrary sql statements against the default database on start up with `spin up --sqlite-migration="YOUR SQL HERE"`

* **I do not believe this feature is ready to marked stable**. As such I have put sqlite support behind a feature flag in the only SDK that currently supports it: Rust. This will still allow folks to target sqlite if they create spin components manually (i.e., using wit-bindgen directly), but I believe we already consider such usage experimental and not officially supported by our backwards compatibility guarantees. If we believe that by merging this PR that we are committing to backwards compatibility of this feature in its current form, then I think we should wait and consider how we want to allow for experimentation before committing to backwards compatibility.
* There are still [some open questions](https://github.com/fermyon/spin/blob/7d97005de534e8516894925dc6c8c177169eeaaf/docs/content/sips/013-sqlite.md?plain=1#L128) in the SIP with regards to sqlite versioning and capacity limits. I don't believe those need to be solved before experimentation can begin.
* There is no support for this feature in Fermyon Cloud. Since this is an experimental feature only, I believe that this is fine for now - though we will quickly want to change that.

File: crates/loader/src/bindle/config.rs
Comment (itowlson) [inline crates/loader/src/bindle/config.rs:52]: I don't know if this counts as human readable text, but if it does, then SUPER DUPER NIT we should follow the project by using "SQLite" rather than "sqlite."

File: docs/content/sips/013-sqlite.md
Comment (itowlson) [inline docs/content/sips/013-sqlite.md:110]: Couple of ill-formed thoughts on this:

* As a point of detail, I suggest `--sqlite` rather than specifically talking about migrations, unless the flag is specifically bound to some sort of migrations format or infrastructure.  It's a pretty general purpose command facility.
* Because SQL statements can be quite long and fiddly, it would be good for users not to have to type them interactively on the CLI.  So it might be worth, even on day one, having a syntax for running a statement (or statements?) from a file, e.g. `--sqlite @usertable.sql`.
* Longer term, could consider a way to connect the user's favourite migration to hook up to the private database so it could generate the required scripts (or execute them directly?).  Not sure what this would look like though.
* `--key-value` currently operates only on the default store.  Are we okay with `--sqlite` operating only on the default database?  I guess if they have non-default ones then those are BYO and they can easily use their favourite SQLite tools on them.

File: docs/content/sips/013-sqlite.md
Comment (itowlson) [inline docs/content/sips/013-sqlite.md:114]: These two lines seem to be talking about something other than the CLI option... am I misunderstanding?

File: docs/content/sips/013-sqlite.md
Comment (itowlson) [inline docs/content/sips/013-sqlite.md:131]: This feels like something we probably don't need to be concerned about in local Spin.  Fermyon Cloud or other hosts may want to place limits, but I don't think we need to plan for that here (except maybe assigning an error code for them).

The main thing I can think of locally is if there could be problems with the size of a query result set, which brings us back round to the list vs cursor vs hybrid issue all over again...

File: docs/content/sips/013-sqlite.md
Comment (itowlson) [inline docs/content/sips/013-sqlite.md:110]: tiny typo 
```suggestion
Database tables typically require some sort of configuration in the form of database migrations to get table schemas into the correct state. To begin with a command line option supplied to `spin up` will be available for running SQL statements on start up (i.e., `--sqlite-migration "CREATE TABLE users..."`). It will be up to the user to provide idempotent statements such that running them multiple times does not produce unexpected results.
```

File: sdk/rust/src/lib.rs
Comment (itowlson) [inline sdk/rust/src/lib.rs:12]: Is this for your testing, or do you intent that SDK users would need to turn the feature on in Cargo.toml until it stabilises?

**Review by itowlson — APPROVED** [review]
Some comments / questions / nits but nothing that can't be iterated on as we build it out.  Nice one - thank you!

File: sdk/rust/src/lib.rs
Comment (itowlson) [reply sdk/rust/src/lib.rs:12]: Oh sorry I see discussion of this in your ready-for-review comment.

**Comment by itowlson** [comment]
By the way, what is the order of landing things for the first iteration?  I saw the PR took a dependency on a branch of `spin-componentize`.  Is the idea that we would land `spin-componentize` first?  Is that going to be the general workflow for as long as we need to continue to support modules?  (cc @dicej)

File: sdk/rust/src/lib.rs
Comment (itowlson) [reply sdk/rust/src/lib.rs:12]: I had a look at the docs (https://developer.fermyon.com/spin/api-guides-overview) and we currently have several APIs that are described as "experimental" but to which don't require a specific opt-in by the developer.  (I don't present this as advocacy for either approach, just as s data point.)

File: sdk/rust/src/sqlite.rs
Comment (itowlson) [inline sdk/rust/src/sqlite.rs:19]: Could we add an `open_default()` please, to save folks having to use the magic string everywhere, and for consistency with the Rust KV SDK?  Thanks!

File: sdk/rust/src/sqlite.rs
Comment (itowlson) [inline sdk/rust/src/sqlite.rs:9]: suggestion: `ParameterValue` would be consistent with our other Rust RDBMS SDKs, and is clear that it's a _value_ rather than a data type.  And similar comment for `DataTypeResult` - the other SDKs use `DbValue` for this.

File: docs/content/sips/013-sqlite.md
Comment (rylev) [reply docs/content/sips/013-sqlite.md:114]: Yes, this is about an alternative approach that we will *not* be doing. I'll try to make the wording more clear.

File: sdk/rust/src/lib.rs
Comment (rylev) [reply sdk/rust/src/lib.rs:12]: The current way of describing things as experimental but not putting any opt-in in front of the user is pretty dangerous, and I think we should move away from doing so. I plan on writing something up on how we can do experimental features in the future.

File: docs/content/sips/013-sqlite.md
Comment (rylev) [reply docs/content/sips/013-sqlite.md:131]: Perhaps we should give some sort of *minimum* requirement that we expect every host environment running spin applications to conform to. The [key-value SIP does this](https://github.com/fermyon/spin/blob/608b5f4b11914109e84f9c950cba6db5b50ce512/docs/content/sips/010-key-value.md?plain=1#L123-L125) for that feature.

File: docs/content/sips/013-sqlite.md
Comment (rylev) [reply docs/content/sips/013-sqlite.md:110]: 1.) I've renamed the option to just `sqlite` 👍
2.) I think this would be a good idea. Is `@` the right syntax for indicating a file?
3 and 4.) I think 3 is related to number 4. If we give the user a way to run sqlite statements (perhaps through a file) on any of their databases, we won't need a different mechanism for the non-default databases.

Comment (radu-matei) [reply docs/content/sips/013-sqlite.md]: Big +1 on something like `--sqlite @usertable.sql`.

File: crates/trigger/src/runtime_config.rs
Comment (dicej) [inline crates/trigger/src/runtime_config.rs:114]: (Optional): could use `HashMap::entry` and https://doc.rust-lang.org/std/collections/hash_map/enum.Entry.html#method.or_insert_with here and above instead of the `contains_key`/`insert` combo.

File: crates/sqlite/src/lib.rs
Comment (dicej) [reply crates/sqlite/src/lib.rs]: I definitely like the `block-in-place` option until we're ready to move to an async sqlite library (which will likely do something similar under the hood).

**Review by dicej — APPROVED** [review]
Looks great overall.

The `async { $BODY }.await` stuff is aggravating my OCD, mainly just because `$BODY` does synchronous I/O and no async, so I'm definitely in favor of using `block_in_place` instead.  No need to split each method, either -- just make it `task::block_in_place(|| { $BODY })`.

File: docs/content/sips/013-sqlite.md
Comment (itowlson) [reply docs/content/sips/013-sqlite.md:136]: I like that plan!

**Review by radu-matei — APPROVED** [review]
LGTM, and the todo example works great for me!

This is really exciting, thanks for working on it, @rylev!

File: crates/trigger/src/runtime_config.rs
Comment (rylev) [reply crates/trigger/src/runtime_config.rs:114]: Entry requires an owned key every time so we would do an extra allocation. I find in this particular case the non-entry based code is easier to read too. I think I'd prefer to keep it as it is, but let me know if you disagree!

File: docs/content/sips/013-sqlite.md
Comment (rylev) [reply docs/content/sips/013-sqlite.md]: Done!

File: docs/content/sips/013-sqlite.md
Comment (rylev) [reply docs/content/sips/013-sqlite.md]: Done!! You can now run sql from files:

```
spin up --sqlite="@migration.sql"
```

**Comment by rylev** [comment]
I think this is ready to merge. We'll wait until the next version of spin is released so that we have a full release cycle to let this bake.

File: docs/content/sips/013-sqlite.md
Comment (lann) [inline docs/content/sips/013-sqlite.md:105]: We could hedge on this by changing the function name to e.g. `query-all`, reserving `query` for a future cursor API.

**Pull request merged**
